### ISSUE TITLE
Bug: [no-floating-promises] allowForKnownSafeCalls isn't allowing react-router's useNavigate

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#10686]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

Given this setting for [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises):

```js
"@typescript-eslint/no-floating-promises": ["error", {
    "allowForKnownSafeCalls": [
        { "from": "package", "name": "useNavigate", "package": "react-router" }
    ]
}]
```

...I would have thought this code would not produce a lint report:

```js
import { useNavigate } from 'react-router';

const navigate = useNavigate();
navigate('/new/page');
```

...but it does on my machine:

```plaintext
/Users/josh/repos/repros/index.ts
  4:1  error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
```

Sparked by a `#help` question in Discord: https://discord.com/channels/1026804805894672454/1326739019387899974

### Reproduction Repository Link

https://github.com/JoshuaKGoldberg/repros/tree/repro-tseslint-no-floating-promises-react-router-allow-useNavigate

### Repro Steps

1) clone the repo at that branch
2) `npm install`
3) `npm run lint`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.20.0` |
| `TypeScript`                           | `5.7.3` |
| `ESLint`                               | `9.18.0` |

**Comment by 43081j** [comment]
This is because `useNavigate` isn't the type, it is the function being called. `allowForKnownSafeCalls` operates on the type, and in this case, that is `NavigateFunction` (since `useNavigate(): NavigateFunction`)

if you set `"name": "NavigateFunction"`, it will work

maybe that's just a docs thing or maybe the rule should operate on symbol names instead

**Comment by JoshuaKGoldberg** [comment]
Aha, thank you! I think this is what @bradzacher mentioned ... somewhere, I forget - that the selector only works on _types_, not _values_, at the moment.

Good question on whether this is a bug or a docs issue. IMO this is a bug because we want to let users specify allowed known safe functions. The specifier is even called _"type or value specifier"_. cc @typescript-eslint/triage-team

**Comment by bradzacher** [comment]
The issue isn't that though.

The promise doesn't come from the `useNavigation` function. It comes from the `navigate` function. That function is locally defined so it will never match the selector you've configured.

Put another way - your config should prevent flagging on
```ts
useNavigation();
```
But not on
```ts
cons cb = useNavigation();
cb(); // error here
```

**Comment by 43081j** [comment]
Though still this option operates on types right now and `useNavigation` isn't that

That is possibly still an issue if we want the rule option to also allow allowing of symbol names rather than only types

**Comment by kirkwaiblinger** [comment]
FYI @JoshuaKGoldberg https://github.com/typescript-eslint/typescript-eslint/issues/10660 also/already tracks matching values with `TypeOrValueSpecifier`. Seemingly, though, since that won't solve this issue, what remains on this issue is more of a "what should be done about patterns that can't easily be matched by a type/value selector?"

**Comment by bradzacher** [comment]
> what should be done about patterns that can't easily be matched by a type/value selector?

Ideally the libraries return some nominally typed promise that a user can match the rule on - which was what we tried to push people to.

Alternately `PromiseLike` is currently ignored by the rule as per other issues because of its existing usages in such APIs.

**Comment by JoshuaKGoldberg** [comment]
Yeah I don't know that there's a way for us to do anything here. As mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/10686#issuecomment-2601042869, targeting the `NavigateFunction` name itself _does_ work in this report:

```js
"@typescript-eslint/no-floating-promises": ["error", {
    "allowForKnownSafeCalls": [
        { "from": "package", "name": "NavigateFunction", "package": "react-router" }
    ]
}]
```

Closing out as non-actionable. Thanks all!
